internal/cluster_reference_test.go

```diff
@@ -81,58 +86,98 @@ var _ = Describe("ParseRabbitmqClusterReference", func() {
 			Expect(svc.ObjectMeta).To(Equal(existingService.ObjectMeta))
 			Expect(svc.Spec).To(Equal(existingService.Spec))
-			Expect(secret.ObjectMeta).To(Equal(existingCredentialSecret.ObjectMeta))
-			Expect(secret.Data).To(Equal(existingCredentialSecret.Data))
+
+			usernameBytes, _ := credsProvider.Data("username")
+			passwordBytes, _ := credsProvider.Data("password")
+			Expect(usernameBytes).To(Equal([]byte(existingRabbitMQUsername)))
+			Expect(passwordBytes).To(Equal([]byte(existingRabbitMQPassword)))
 		})
 
-		When("RabbitmqCluster does not have status.binding set", func() {
+		When("RabbitmqCluster does not have status.defaultUser set", func() {
 			BeforeEach(func() {
 				*existingRabbitMQCluster = rabbitmqv1beta1.RabbitmqCluster{
 					ObjectMeta: metav1.ObjectMeta{
 						Name:      "rmq-incomplete",
 						Namespace: "rabbitmq-system",
 					},
 					Status: rabbitmqv1beta1.RabbitmqClusterStatus{
-						DefaultUser: &rabbitmqv1beta1.RabbitmqClusterDefaultUser{
-							ServiceReference: &rabbitmqv1beta1.RabbitmqClusterServiceReference{
-								Name:      "rmq-service",
-								Namespace: "rabbitmq-system",
-							},
+						Binding: &corev1.LocalObjectReference{
+							Name: "rmq-default-user-credentials",
 						},
 					},
 				}
 			})
 
 			It("errors", func() {
 				_, _, _, err := internal.ParseRabbitmqClusterReference(ctx, fakeClient, topology.RabbitmqClusterReference{Name: existingRabbitMQCluster.Name}, existingRabbitMQCluster.Namespace)
-				Expect(err).To(MatchError("no status.binding set"))
+				Expect(err).To(MatchError("no status.defaultUser set"))
 			})
 		})
 
-		When("RabbitmqCluster does not have status.defaultUser set", func() {
+		When("vault secret backend is declared on cluster spec", func() {
+			var (
+				err                   error
+				fakeSecretStoreClient *internalfakes.FakeSecretStoreClient
+				credsProv             internal.CredentialsProvider
+			)
+
 			BeforeEach(func() {
 				*existingRabbitMQCluster = rabbitmqv1beta1.RabbitmqCluster{
 					ObjectMeta: metav1.ObjectMeta{
-						Name:      "rmq-incomplete",
+						Name:      "rmq",
 						Namespace: "rabbitmq-system",
 					},
 					Status: rabbitmqv1beta1.RabbitmqClusterStatus{
 						Binding: &corev1.LocalObjectReference{
 							Name: "rmq-default-user-credentials",
 						},
+						DefaultUser: &rabbitmqv1beta1.RabbitmqClusterDefaultUser{
+							ServiceReference: &rabbitmqv1beta1.RabbitmqClusterServiceReference{
+								Name:      "rmq",
+								Namespace: "rabbitmq-system",
+							},
+						},
+					},
+					Spec: rabbitmqv1beta1.RabbitmqClusterSpec{
+						SecretBackend: rabbitmqv1beta1.SecretBackend{
+							Vault: &rabbitmqv1beta1.VaultSpec{
+								Role:            "sausage",
+								DefaultUserPath: "/some/path",
+							},
+						},
 					},
 				}
+
+				fakeSecretStoreClient = &internalfakes.FakeSecretStoreClient{}
+				fakeSecretStoreClient.ReadCredentialsReturns(fakeCredentialsProvider, nil)
```

Comment:
@georgeharley When I run `make unit-tests` this test fails because `fakeCredentialsProvider` is null at this line which means the `fakeSecretStoreClient` will return a null CredentialsProvider and test fails. This `beforeEach` method is invoked before `JustBeforeEach` method of the `Describe`. 
The behaviour of `JustBeforeEach` is explained [here](https://onsi.github.io/ginkgo/#separating-creation-and-configuration-justbeforeeach). It has also trapped me a few times.
 
One easy way to fix it is by moving the initialization of the `fakeCredentialsProvider` to the `BeforeEach` of this test case.

Reply:
When `make unit-tests` is run in my environment there doesn't appear to be any failing tests. Could be something weird in my setup or else the randomising of the test specs is somehow always falling in my favour? FWIW I'm using go version 1.17.2, Ginkgo version 1.16.5, and GNU Make version 3.8.1 all running on macOS Monterey 12.0.1.

I'll make the suggested fix to ensure the outcome is predictable.

Reply:
Test updated.